src/spring-cloud/azext_spring_cloud/_buildservices_factory.py

```diff
@@ -0,0 +1,197 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+# pylint: disable=too-few-public-methods, unused-argument, redefined-builtin
+import sys
+import requests
+from time import sleep
+from requests.auth import HTTPBasicAuth
+from knack.log import get_logger
+from azure.cli.core.azclierror import InvalidArgumentValueError
+from msrestazure.tools import parse_resource_id
+from azure.cli.core.commands.client_factory import get_subscription_id
+from msrestazure.azure_exceptions import CloudError
+from knack.util import CLIError
+from .vendored_sdks.appplatform.v2022_01_01_preview import models
+from ._deployment_uploadable_factory import uploader_selector
+from ._log_stream import LogStream
+
+logger = get_logger(__name__)
+
+
+class BuildService:
+    def __init__(self, cmd, client, resource_group, service):
+        self.cmd = cmd
+        self.client = client
+        self.resource_group = resource_group
+        self.service = service
+        self.name = 'default'
+        self.log_stream = None
+        self.progress_bar = None
+        self.terminated_state = ['Succeeded', 'Failed', 'Deleting']
+
+    def get_total_steps(self):
+        return 4
+
+    def build_and_get_result(self, total_steps, **kwargs):
+        logger.warning("[1/{}] Requesting for upload URL.".format(total_steps))
+        upload_info = self._get_upload_info()
+        logger.warning("[2/{}] Uploading package to blob.".format(total_steps))
+        uploader_selector(upload_url=upload_info.upload_url, **kwargs).upload_and_build(**kwargs)
+        logger.warning("[3/{}] Creating or Updating build '{}'.".format(total_steps, kwargs['app']))
+        build_result_id = self._queue_build(upload_info.relative_path, **kwargs)
+        logger.warning("[4/{}] Waiting for building docker image to finish. This may take a few minutes.".format(total_steps))
+        self._wait_build_finished(build_result_id)
+        return build_result_id
+
+    def _get_upload_info(self):
+        try:
+            response = self.client.build_service.get_resource_upload_url(self.resource_group, self.service, self.name)
+            if not response.upload_url:
+                raise CLIError("Failed to get a SAS URL to upload context.")
+            return response
+        except CloudError as e:
+            raise CLIError("Failed to get a SAS URL to upload context. Error: {}".format(e.message))
+        except AttributeError as e:
+            raise CLIError("Failed to get a SAS URL to upload context. Error: {}".format(e))
```

Comment:
Please use the specific error type instead of `CLIError`

Reply:
Updated